app/src/main/java/com/github/polybooks/core/databaseImpl/OLBookDatabase.kt

```diff
@@ -0,0 +1,230 @@
+
+
+private const val TITLE_FIELD_NAME = "full_title"
+private const val AUTHORS_FIELD_NAME = "authors"
+private const val FORMAT_FIELD_NAME = "physical_format"
+private const val ISBN13_FIELD_NAME = "isbn_13"
+private const val PUBLISHER_FIELD_NAME = "publishers"
+private const val PUBLISH_DATE_FIELD_NAME = "publish_date"
+
+private const val DATE_FORMAT = "MMM dd, yyyy"
+private const val ISBN13_FORMAT = """[0-9]{13}"""
+private const val ISBN10_FORMAT = """[0-9]{9}[0-9X]"""
+private const val ISBN_FORMAT = """($ISBN10_FORMAT)|($ISBN13_FORMAT)"""
+
+private const val OL_BASE_ADDR = """https://openlibrary.org"""
+
+/**
+ * An implementation of a book database based on the Open Library online database
+ * */
+@SuppressLint("NewApi")
+class OLBookDatabase : BookDatabase {
+
+    override fun queryBooks(): BookQuery = OLBookQuery()
+
+    private inner class OLBookQuery : BookQuery {
+
+        private var ordering = DEFAULT
+
+        private var empty : Boolean = false
+        private var title : String? = null
+        private var isbn : String? = null
+
+        override fun onlyIncludeInterests(interests: Collection<Interest>): BookQuery {
+            System.err.println("Warning: onlyIncludeInterest not fully implemented for OLBookQuery")
+            this.empty = true
+            return this
+        }
+
+        override fun searchByTitle(title: String): BookQuery {
+            this.empty = false
+            this.title = title
+            this.isbn = null
+            return this
+        }
+
+        override fun searchByISBN13(isbn13: String): BookQuery {
+            this.empty = false
+            this.title = null
+            this.isbn = isbn13
+            return this
+        }
+
+        override fun withOrdering(ordering: BookOrdering): BookQuery {
+            this.ordering = ordering
+            return this
+        }
+
+        override fun getAll(): CompletableFuture<List<Book>> {
+            if (empty) return CompletableFuture.completedFuture(Collections.emptyList())
+            else if (title != null) {
+                return TODO("Search by title")
+            } else { assert(isbn != null)
+                val url = userISBN2URL(isbn!!)
+                if (url != null) {
+                    return url2json(url)
+                        .thenApplyAsync { listOf(parseBook(it)) }
+                        .exceptionally { exception ->
+                            if (exception is CompletionException && exception.cause is FileNotFoundException)
+                                return@exceptionally Collections.emptyList<Book>()
+                            else if (exception is CompletionException) throw exception.cause!!
+                            else throw exception
+                    }
+                } else {
+                    val future = CompletableFuture<List<Book>>()
+                    future.completeExceptionally(Exception("ISBN is not valid"))
+                    return future
```

Comment:
When SaleDb will be merged, we will be able to use DatabaseExceptions instead. Might be useful later for activities using our db interfaces

Reply:
Yup, I saw it, it was a good idea to make a DatabaseException

Reply:
done.